text/0000-cargo-alternative-registry-auth.md

```diff
@@ -0,0 +1,107 @@
+
+There are many multi-protocol package managers: Artifactory, AWS CodeArtifact, Azure Artifacts, GitHub Artifacts, Google Artifact Registry, and CloudSmith. However, only CloudSmith and Artifactory support Cargo, and they resort to encoding credentials in the URL or allowing anonymous download of packages. This RFC (especially when combined with the approved http-registry RFC) will make it significantly easier to implement Cargo support on private package managers.
+
+# Guide-level explanation
+Alternative registry operators can set a new key `auth-required = true` in the registry's `config.json` file, which will cause Cargo to include the Authorization token for all API requests, crate downloads, and index updates (if over HTTP).
+
+```json
+{
+    "dl": "https://example.com/index/api/v1/crates",
+    "api": "https://example.com/",
+    "auth-required": true
+}
+```
+
+If the index is hosted via HTTP using [RFC2789](https://github.com/rust-lang/rfcs/pull/2789) and Cargo receives an `HTTP 401` error when fetching `config.json`, Cargo will automatically re-try the request with the Authorization token included.
+
+
+# Reference-level explanation
+A new key, `auth-required`, will be allowed in the [`config.json`](https://doc.rust-lang.org/cargo/reference/registries.html#index-format) file stored in the registry index. When this key is set to `true`, the authorization token will be sent with any HTTP requests made to the registry API, crate downloads, and index (if using http). If a token is not available when Cargo is attempting to make a request, the user would be prompted to run `cargo login --registry NAME` to save a token.
+
+The authorization token would be sent as an HTTP header, exactly how it is currently sent for operations such as `publish` or `yank`:
+```
+Authorization: <token>
+```
+
+## Interaction with HTTP registries
+The approved (but currently unimplemeneted) [RFC2789](https://github.com/rust-lang/rfcs/pull/2789) enables Cargo to fetch the index over HTTP. When fetching `config.json` from an HTTP index, if Cargo receives an `HTTP 401` response, the request will be re-attempted with the Authorization header included. If no authorization token is available, Cargo will suggest that the user run `cargo login` to add one.
+
+To avoid the overhead of an extra HTTP request when fetching `config.json`, the user can optionally configure Cargo locally by setting `auth-required = true` in the `[registries]` table. If the local `auth-required` flag is `true` then Cargo will always include the Authorization token fetching `config.json` over HTTP -- skipping the initial unauthorized requiest and `HTTP 401`. The local configuration option does not impact other operations, such as API requests or downloads. It also does not impact git-based registries.

```

Comment:
Small typo:

```suggestion
To avoid the overhead of an extra HTTP request when fetching `config.json`, the user can optionally configure Cargo locally by setting `auth-required = true` in the `[registries]` table. If the local `auth-required` flag is `true` then Cargo will always include the Authorization token fetching `config.json` over HTTP -- skipping the initial unauthorized request and `HTTP 401`. The local configuration option does not impact other operations, such as API requests or downloads. It also does not impact git-based registries.
```

Also, can you clarify this part:

> The local configuration option does not impact other operations, such as API requests or downloads. It also does not impact git-based registries.

Does this mean that if the local configuration option has `auth-required = true` in the registries table, but the registry has `auth-required = false` in the `config.json` file, then API requests and downloads will *not* include an authorization header?

If so, is there a potential confusion hazard by having two keys with the same name that do slightly different things?

Reply:
Thanks for catching the typo!

I agree there's a potential confusion hazard between the two keys. Only registry implementations/owners will be interested in the key in `config.json` while consumers will be interested in the one in the `[registries]` table.

One solution would be to change the key in `config.json` to something like:
`authorize = 'all'`. Having a string here may allow more flexibility if we want to allow registries to more selectively require authorization. Or we can keep it as a boolean and just pick a different name. I'm open to suggestions :)

Reply:
I was just looking over this RFC with @mmarston senior dev on AWS CodeArtifact, and he was also thrown off by the two different configurations by the same name. After explaining that they where different and what each one did, he was surprised that if a users sets `auth-required = true` that cargo would ever not send the header.  I explained that we wanted to reduce the number of urls where cargo sends the secret, without the server asking for it (in case the server is not going to handle it correctly by say logging it). Over all it is a user experience issue. We came up with 3 ways forward:
1. Make the user setting controlling. This may send creds where thy are not wanted, but if that is what the user asked for then maybe it is ok.
2. Make the user setting more descriptive, auth-on-first-contact = true or auth-required = 'config.json' or suggestions welcome.
3. Save "avoid the one round trip" for later and move the user setting to Future possibilities .

I know I am the one that was concerned about the overhead, but I now wonder if there are users that really care. If no one cares, then lets leave it for a follow up RFC.

Reply:
Just a simple thought about this three points:

Using a string for the `auth-required ` key might give the user more flexibility.

For example valid values might be

-  `always` for always sending the token
- `if-needed` for sending the token only In case the server response with 401
- `never` as default

Reply:
>  I explained that we wanted to reduce the number of urls where cargo sends the secret, without the server asking for it (in case the server is not going to handle it correctly by say logging it)

This seems like it has a few defense-in-depth benefits for the fringe case of a client-side misconfiguration.

However, if the token is appropriately scoped and has a short TTL, I think it's a lot more confusing not to always send it:

- An appropriately scoped token should only confer authority to the server it's being sent to. If that server is in some way leaking the token, it's leaking authority to itself, which makes any compromise only as severe as any other breakage in itself
- It will be a lot more obvious the server is logging the token in bad places if it's logged on every request, as opposed to specifically authenticated actions

Reply:
> However, if the token is appropriately scoped and has a short TTL, I think it's a lot more confusing not to always send it

It feels like too large of a breaking change to unconditionally send the token with all requests. If we were designing the private registry feature from the ground up here it would be different.

> I know I am the one that was concerned about the overhead, but I now wonder if there are users that really care. If no one cares, then lets leave it for a follow up RFC.

Agreed. I think removing the local configuration option makes the most sense. It prevents local misconfiguration and a single extra HTTP request per cargo invocation should have minimal performance impact. As you said, we can always add a follow up RFC if necessary.

I'll update this RFC to move this to the Future Possibilities section.